src/Storages/ObjectStorage/StorageObjectStorageSource.cpp

```diff
@@ -494,6 +512,128 @@ StorageObjectStorageSource::ReaderHolder StorageObjectStorageSource::createReade
         return schema_cache->tryGetNumRows(cache_key, get_last_mod_time);
     };
 
+    /// List of columns with constant value in current file, and values
+    std::map<size_t, ConstColumnWithValue> constant_columns_with_values;
+    std::unordered_set<String> constant_columns;
+
+    NamesAndTypesList requested_columns_copy = read_from_format_info.requested_columns;
+
+    std::unordered_map<String, std::pair<size_t, NameAndTypePair>> requested_columns_list;
+    {
+        size_t column_index = 0;
+        for (const auto & column : requested_columns_copy)
+            requested_columns_list[column.getNameInStorage()] = std::make_pair(column_index++, column);
+    }
+
+    if (context_->getSettingsRef()[Setting::allow_experimental_iceberg_read_optimization])
+    {
+        auto file_meta_data = object_info->getFileMetaInfo();
+        if (file_meta_data.has_value())
```

Comment:
again, this is an excellent example: `file_meta_data.value()` can be null here. I know that there are precautions now, but this is fragile, IMO simple `std::shared_ptr` would be more robust here.
```suggestion
        if (file_meta_data)
```
and 
```
file_meta_data.value()->columns_info
```
would be replaced by just 
```
file_meta_data->columns_info
```